packages/js/src/utils/generators.test.ts

```diff
@@ -0,0 +1,77 @@
+import { UrlSafeBase64 } from '@silverhand/essentials';
+import { toUint8Array } from 'js-base64';
+
+import { generateCodeChallenge, generateCodeVerifier, generateState } from './generators';
+
+const isUrlSafe = (input: string) => /^[\w-]*$/.test(input);
+
+describe('generateState', () => {
+  test('should be random value', () => {
+    const state1 = generateState();
+    const state2 = generateState();
+    expect(state1).not.toEqual(state2);
+  });
+
+  test('should be url-safe', () => {
+    const state = generateState();
+    expect(isUrlSafe(state)).toBeTruthy();
+  });
+
+  test('raw random data length should be length 64', () => {
+    const state = generateState();
+    expect(toUint8Array(UrlSafeBase64.restoreNonUrlSafeCharacters(state)).length).toEqual(64);
+  });
+});
+
+describe('generateCodeVerifier', () => {
+  test('should be random value', () => {
+    const codeVerifier1 = generateCodeVerifier();
+    const codeVerifier2 = generateCodeVerifier();
+    expect(codeVerifier1).not.toEqual(codeVerifier2);
+  });
+
+  test('should be url-safe', () => {
+    const codeVerifier = generateCodeVerifier();
+    expect(isUrlSafe(codeVerifier)).toBeTruthy();
+  });
+
+  test('raw random data length should be length 64', () => {
+    const codeVerifier = generateCodeVerifier();
+    expect(toUint8Array(UrlSafeBase64.restoreNonUrlSafeCharacters(codeVerifier)).length).toEqual(
+      64
+    );
+  });
+});
+
+describe('generateCodeChallenge', () => {
+  test('dealing with different code verifiers should not be equal', async () => {
+    const codeVerifier1 =
+      '1LugF_nxSLUElpAECudk0wDYjfkJL54FgzZtGtfe0Q6-ytCA5DUGRwKBzKHfB7tFtdyQ1GxFLBGF3Da4SxrAAA';
+    const codeVerifier2 =
+      'npz_GJ1zmSQFVXXEGWaaUXln5uekO9DlKTTXPL_motMEhRW-SG7YKMZiKYa78197UBmL7avtz-uMUWlUp2__Gw';
+
+    const codeChallenge1 = await generateCodeChallenge(codeVerifier1);
+    const codeChallenge2 = await generateCodeChallenge(codeVerifier2);
+    expect(codeChallenge1).not.toEqual(codeChallenge2);
+  });
+
+  test('dealing with same code verifier should be equal', async () => {
+    const codeVerifier =
+      'KH5wVipNo1W0cpKnx0csh_WlmY5fpmAiYl46OBBc2SE6j93RtKo8j7M230jQ9DF0IWBFl1j_UsrI_5dcpHdo1w';
+    const codeChallenge1 = await generateCodeChallenge(codeVerifier);
+    const codeChallenge2 = await generateCodeChallenge(codeVerifier);
+    expect(codeChallenge1).toEqual(codeChallenge2);
+  });
+
+  test('dealing with code verifier containing latin1 character', async () => {
```

Comment:
Discuss: is it necessary test cases?
(added according to [unit tests](https://github.com/logto-io/swift/blob/master/Tests/LogtoTests/LogtoUtilitiesTests.swift#L18-L38) from Swift SDK)

Reply:
@logto-io/eng any idea?